### PR TITLE
Add PKCE support when Authorization Code flow is used

### LINKED ISSUE #5348: Support OAuth 2.0 Authorization Code flow with PKCE
### Is your feature request related to a problem?
In the light of the new IETF related to [OAuth 2.0 for Browser-Based Apps](https://tools.ietf.org/html/draft-parecki-oauth-browser-based-apps-02#page-3) (see 4. Overview), an application running in a browser and using the Authorization Code flow should be using PKCE ([Proof Key for Code Exchange](https://tools.ietf.org/html/rfc7636)) for increased security. It is not the case currently with swagger-ui. This [article](https://brockallen.com/2019/01/03/the-state-of-the-implicit-flow-in-oauth2/) by Brock Allen also touches on the subject.

### Describe the solution you'd like
Everytime the `authorizationCode` flow is used (these are based on the PKCE IETF linked above):
1. A `code_verifier` is generated (potentially using `uuid4` like they do in [oidc-client-js](https://github.com/IdentityModel/oidc-client-js)). This value needs to stored somewhere because it needs to be sent on the Token request (looking at the code, I figured `win.swaggerUIRedirectOauth2` would be the right place to store this).
2. Create the `code_challenge` for the `code_verifier` (hashed using SHA256) and send this on the Authorize request alongside the `code_challenge_method`.

### Describe alternatives you've considered
- I've considered using `additionalQueryStringParams` to feed in the `code_verifier` and the `code_challenge`, but these values need to be generated on each authorize request.
- I've considered using the Inject JavaScript functionality, but I feel like I would be replacing whole methods from swagger-ui which would bind me to a certain version of the code.

### Additional context
I feel this is something that should be introduced into swagger-ui because of the added security it provides. Everybody would and *should* benefit from it.

### PR DESCRIPTION
**Pull request by poveilleux**
### Description
- Added a new `usePkce` flag that can be enabled to use [PKCE](https://tools.ietf.org/html/rfc7636) with the Autorization Code flow. This flag is used when the `/authorize` request is sent (`oauth2authorize.js`) to determine if the `code_challenge` has to be sent (part of the PKCE protocol). The `code_challenge` is a SHA256 of the `code_verifier` which is sent on the `/token` request.
- When the `code_challenge` is sent, the `code_verifier` is stored in the `auth` object to be sent on the `/token` request.
- The `code_verifier` needs to have a length of at least 43 characters as per the spec.

### Motivation and Context
- This makes the Authorization Code flow more secure because the `code` can only be used by the client that provided the `code_challenge` on the `/authorize` request.
- Fixes #5348 

### How Has This Been Tested?
My setup:
1. An OpenID Connect compliant STS using IdentityServer4. I configured a test `Client` that had `RequirePkce` enabled with the Authorization Code flow. With these settings, it was not possible to use the Code flow without sending the required data for PKCE.
2. An API using Swashbuckle.AspNetCore and ASP.Net Core to generate a `swagger.json` configured to use my STS.

How I tested and implemented the fix:
1. I set up the dev-server to use the `swagger.json` from my API. To be able to request an access token, I needed to have PKCE working.
2. I made the "Authorize" request from SwaggerUI. I was getting an error on the `/authorize` request about the missing `code_challenge`. I made the required changes to make that part work.
3. I tested again and I got an error on the `/token` request. I made the required changes to send the `code_verifier` and it worked (meaning I got the access token to make requests to the API).

Impact on the rest of the code:
- None, the use of PKCE is opt-in with a `usePkce` flag on the `authConfigs`. This way, it is not a breaking change for people wanting to upgrade to this new version of SwaggerUI.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [x] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by poveilleux** [comment]
This is weird. Locally I get errors in `test/e2e-cypress/tests/features/oauth2-flows/application.js` and `test/e2e-cypress/tests/features/oauth2-flows/password.js` but they run just fine in Jenkins.

I am also getting an error on the `test/core/system/system.jsx` about `render()` that should have a return statement (eslint error) but I did not change this.

**Comment by lsivakumar** [comment]
Hi, Is this getting merged to master soon ?

**Comment by poveilleux** [comment]
@lsivakumar I'd like to, but as you see, the build fails for a reason I don't really understand (I didn't change those files). I am not sure how I can poke the people maintaining SwaggerUI.

**Comment by shockey** [comment]
@poveilleux, thank you for your work here!

I received an email about this PR, thought it'd be helpful to copy here:

> > Can [this PR] be merged to master?
>
> Hi!
> 
> We're currently sitting on a big backlog of PRs that we need to work through, but rest assured that we haven't forgotten about this!

**Comment by poveilleux** [comment]
Thanks @shockey for the feedback. At least I know it was not forgotten!

**Comment by poveilleux** [comment]
Hi @shockey, is there anything I can do to help you pushing this pull request into master faster? Thanks

**Comment by fuzzzerd** [comment]
I too would love to see this merged so work on [adding it downstream](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/999) can begin.

**Comment by fuzzzerd** [comment]
Are there plans to merge this? If so, when?

**Comment by poveilleux** [comment]
@shockey So, it's been 4 months since I opened this PR. I keep fixing the merge conflicts so you guys can merge the PR when you get the time and it does not take too long. Is there anything else I can do to help you to get these changes in master faster? Since the use of PKCE with the Authorization Code flow is the new security recommendation, I think more interest should be put into this pull request.

Thank you

File: src/core/oauth2-authorize.js [hunk 1]
```diff
@@ -66,6 +68,22 @@ export default function authorize ( { auth, authActions, errActions, configs, au
     query.push("realm=" + encodeURIComponent(authConfigs.realm))
   }
 
+  if (isCode && authConfigs.usePkceWithAuthorizationCodeGrant) {
```
Comment (shockey) [inline src/core/oauth2-authorize.js]: **Change request**: instead of setting and using `isCode`, let's use `flow === "authorizationCode"`, unless there's a reason not to that I'm missing.

File: src/core/plugins/auth/actions.js [hunk 2]
```diff
@@ -120,13 +120,14 @@ export const authorizeApplication = ( auth ) => ( { authActions } ) => {
...
-  let { schema, name, clientId, clientSecret } = auth
+  let { schema, name, clientId, clientSecret, codeVerifier } = auth
   let form = {
     grant_type: "authorization_code",
     code: auth.code,
     client_id: clientId,
     client_secret: clientSecret,
-    redirect_uri: redirectUrl
+    redirect_uri: redirectUrl,
+    code_verifier: codeVerifier
```
Comment (shockey) [inline src/core/plugins/auth/actions.js]: **Question**: does this omit `code_verifier` completely when `codeVerifier` is undefined?

File: src/core/oauth2-authorize.js [hunk 3]
```diff
@@ -66,6 +68,22 @@ export default function authorize ( { auth, authActions, errActions, configs, au
...
+      // as per the PKCE spec, the Code Verifier must have a minumum length of 43 characters
+      // and random() produces a value that has a length of 32 characters
+      // source: https://tools.ietf.org/html/rfc7636#section-4.1
+      const codeVerifier = random() + random() + random()
```
Comment (shockey) [inline src/core/oauth2-authorize.js]: **Question**: is there a reason we're relying on a UUID library for this?

From § 4.1:

> It is RECOMMENDED that the output of
>  a suitable random number generator be used to create a 32-octet
>  sequence.  The octet sequence is then base64url-encoded to produce a
>  43-octet URL safe string to use as the code verifier.

Seems to me we could more easily do something like:

```js
// adapted from https://auth0.com/docs/flows/guides/auth-code-pkce/includes/create-code-verifier
function generateCodeVerifier() {
  return crypto.randomBytes(32)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}
```

(Our Webpack config lets us use Node `crypto`, AFAIK)

**Comment by poveilleux** [comment]
@shockey Thank you, I am working on the requested changes right now.

File: src/core/oauth2-authorize.js (hunk 1 above)
Comment (poveilleux) [reply src/core/oauth2-authorize.js]: No particular reason at all. I thought I would reuse the fact this condition has been already evaluated. I changed it to your suggestion.

File: src/core/oauth2-authorize.js (hunk 3 above)
Comment (poveilleux) [reply src/core/oauth2-authorize.js]: I like what you suggested better. I added a function at the end of the file and called it here.

File: src/core/plugins/auth/actions.js (hunk 2 above)
Comment (poveilleux) [reply src/core/plugins/auth/actions.js]: Yes it will. That's what `buildFormData` does (skipping undefined values).

File: src/core/utils.js [hunk 4]
```diff
@@ -859,3 +861,15 @@ export function paramToValue(param, paramValues) {
...
+  return uuid4().replace(/-/g, "")
+}
+
+export function hashString(value, alg) {
+    return crypto.Util.hashString(value, alg)
+}
+
+export function hexToBase64Url(value) {
+    return hextob64u(value)
+}
```
[the remark this answers is not in the document]
Comment (poveilleux) [reply src/core/utils.js]: I agree. It is done now.

File: test/mocha/core/oauth2-authorize.js [hunk 5]
```diff
@@ -35,5 +35,25 @@ describe("oauth2", function () {
...
+      const expectedCodeChallenge = "mock_code_challenge"
+      utils.hexToBase64Url = createSpy().andReturn(expectedCodeChallenge)
+
+      authConfig.authConfigs.usePkceWithAuthorizationCodeGrant = true
+
+      oauth2Authorize(authConfig)
+      expect(win.open.calls.length).toEqual(1)
+
+      const actualUrl = new URLSearchParams(win.open.calls[0].arguments[0])
+      expect(actualUrl.get("code_challenge")).toBe(expectedCodeChallenge)
```
[the remark this answers is not in the document]
Comment (poveilleux) [reply test/mocha/core/oauth2-authorize.js]: I applied the spy strategy to the library directly.

**Comment by poveilleux** [comment]
@shockey Everything should be good now. Thank you for taking the time to review the pull request!

File: src/core/utils.js [hunk 6]
```diff
@@ -859,3 +861,19 @@ export function paramToValue(param, paramValues) {
 
   return values[0]
 }
+
+// adapted from https://auth0.com/docs/flows/guides/auth-code-pkce/includes/create-code-verifier
+export function generateCodeVerifier() {
```
Comment (poveilleux) [inline src/core/utils.js]: After realizing my code coverage for the feature was not the best, I moved the 2 methods to create the `code_verifier` and `code_challenge` in the utils file. This enables to mock properly when testing oauth2-authorize.js.

File: test/mocha/core/utils.js [hunk 7]
```diff
@@ -1402,4 +1404,22 @@ describe("utils", function() {
...
+    it("should generate a value of at least 43 characters", () => {
+      const codeVerifier = generateCodeVerifier()
+
+      // Source: https://tools.ietf.org/html/rfc7636#section-4.1
+      expect(codeVerifier.length).toBeGreaterThanOrEqualTo(43)
+    })
+  })
+
+  describe("createCodeChallenge", function() {
+    it("should hash the input using SHA256 and output the base64 url encoded value", () => {
```
Comment (poveilleux) [inline test/mocha/core/utils.js]: I know hard-coding the 2 values is not the best, but I wanted to make sure the `code_challenge` was indeed the SHA256 of the `code_verifier` in base64 url format.

File: test/mocha/core/oauth2-authorize.js [hunk 8]
```diff
@@ -20,6 +20,18 @@ describe("oauth2", function () {
   }
 
   describe("authorize redirect", function () {
+    let initialGenerateCodeVerifier,
+        initialCreateCodeChallenge
+
+    beforeEach(() => {
```
Comment (poveilleux) [inline test/mocha/core/oauth2-authorize.js]: Not doing resulted in the methods being mocked for all the other tests. Do you have any idea how I could do it differently?

File: test/mocha/core/utils.js (hunk 7 above)
Comment (shockey) [reply test/mocha/core/utils.js]: No problem, but let's add a comment next to the values so it's clear what they are?

File: test/mocha/core/oauth2-authorize.js (hunk 8 above)
Comment (shockey) [reply test/mocha/core/oauth2-authorize.js]: Yes! `expect` has a [`spyOn(target, "method")` function](https://github.com/mjackson/expect#spyon).

I've switched this PR over to it in 898bfc26b.

File: src/core/utils.js [hunk 9]
```diff
@@ -22,6 +22,8 @@ import { memoizedSampleFromSchema, memoizedCreateXMLExample } from "core/plugins
 import win from "./window"
 import cssEscape from "css.escape"
 import getParameterSchema from "../helpers/get-parameter-schema"
+import crypto from "crypto"
+import { hextob64u } from "jsrsasign"
```
Comment (shockey) [inline src/core/utils.js]: **Change request**: `jsrsasign` is [a ~250KB library](https://bundlephobia.com/result?p=jsrsasign@8.0.12), which is quite big. Can we do something lighter here?

File: test/mocha/core/utils.js [hunk 10]
```diff
@@ -1402,4 +1404,22 @@ describe("utils", function() {
...
+      const codeVerifier = "cY8OJ9MKvZ7hxQeIyRYD7KFmKA5znSFJ2rELysvy2UI"
+      const expectedCodeChallenge = "LD9lx2p2PbvGkojuJy7-Elex7RnckzmqR7oIXjd4u84"
```
Comment (shockey) [inline test/mocha/core/utils.js]: **Change request**: add a comment explaining what these values represent

File: test/mocha/core/oauth2-authorize.js (hunk 8 above)
Comment (poveilleux) [reply test/mocha/core/oauth2-authorize.js]: That is nice, I did not know about it! Thanks for making the change.

File: test/mocha/core/utils.js (hunk 10 above)
Comment (poveilleux) [reply test/mocha/core/utils.js]: Done!

File: src/core/utils.js (hunk 9 above)
Comment (poveilleux) [reply src/core/utils.js]: If I knew it was that big, I would not have included it to the dependencies. I added a method to convert base64 string to base64url-encoded, which enabled me to get rid of jsrsasign completely.

**Comment by shockey** [comment]
merged! thanks for the work here, @poveilleux 🙂

**Comment by poveilleux** [comment]
Thanks @shockey for merging the pull request! 😊

**Comment by vgiannone3** [comment]
Hello! I am curious on the status of this request. When will it be available for use? I would like to secure my apps with this technique ASAP. Thanks!

**Comment by webron** [comment]
@vgiannone3 it should already be available in the latest version.
